### PR TITLE
Fault Tolerant Manual: utilities cleanup

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR re-organizes some utilities and remove dead code.

Find the PoC here: #10568

Parts of #10431

To be merged after #10699

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -394,52 +394,6 @@ def _cycle_to_next_worker_and_reset(dataloader: DataLoader, state_dict: Dict[str
     return iter_dataloader
 
 
-def _dataloader_to_state_dict(
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:397]: This are dead utilities. They aren't being used anymore.

File: tests/utilities/test_auto_restart.py [hunk 2]
```diff
@@ -665,44 +663,6 @@ def create_iterable_dataset(batch_size, num_workers, attr_name="iter_sampler", w
     return dataset
 
 
-def test_dataloader_to_state_dict_and_reload():
```
Comment (tchaton) [inline tests/utilities/test_auto_restart.py:668]: Note reviewer: Drop dead code and associated test.

**Review by awaelchli — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: pytorch_lightning/utilities/auto_restart.py [hunk 3]
```diff
@@ -476,6 +430,48 @@ def _capture_metadata_collate(
     return {"data": data, AutoRestartBatchKeys.PL_RESTART_META: metadata}
 
 
+# TODO: Merge this code within stateful DataLoaderIter.
+def _next_data_wrapper(fn, it, dl, num_batches_fetched, data_fetcher) -> Callable:
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:434]: typing annotations would be nice

File: pytorch_lightning/utilities/auto_restart.py [hunk 4]
```diff
@@ -476,6 +430,48 @@ def _capture_metadata_collate(
...
+    @wraps(fn)
+    def wrapper():
+        nonlocal num_batches_fetched
+        nonlocal it
+        nonlocal dl
+
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:440]: These serve no purpose now afaik

```suggestion
```

num_batches_fetched is not used anywhere else outside of the wrapper and it and dl would be modified by reference anyways

File: pytorch_lightning/utilities/auto_restart.py [hunk 5]
```diff
@@ -476,6 +430,48 @@ def _capture_metadata_collate(
     return {"data": data, AutoRestartBatchKeys.PL_RESTART_META: metadata}
 
 
+# TODO: Merge this code within stateful DataLoaderIter.
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:433]: curious: is this something you plan to do shortly or a "would be nice"

File: pytorch_lightning/utilities/auto_restart.py [hunk 6]
```diff
@@ -476,6 +430,48 @@ def _capture_metadata_collate(
...
+def _next_data_wrapper(fn, it, dl, num_batches_fetched, data_fetcher) -> Callable:
+    @wraps(fn)
+    def wrapper():
+        nonlocal num_batches_fetched
+        nonlocal it
+        nonlocal dl
+
+        dataset = dl.dataset
+        combined_batch = fn()
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:442]: I don't get why this is a wrapper if calling the passed fn is the first thing done inside.

File: pytorch_lightning/utilities/data.py [hunk 7]
```diff
@@ -305,3 +297,15 @@ def _dataloader_init_kwargs_resolve_sampler(
 def _auto_add_worker_init_fn(dataloader: DataLoader, rank: int) -> None:
     if int(os.environ.get("PL_SEED_WORKERS", 0)) and dataloader.worker_init_fn is None:
         dataloader.worker_init_fn = partial(pl_worker_init_function, rank=rank)
+
+
+def _apply_fault_tolerant_automatic_capture_dataset_wrapper(dl_kwargs: Dict) -> Dict:
+    dataset = dl_kwargs["dataset"]
+    if isinstance(dataset, IterableDataset):
+        # wrap the `IterableDataset` into a `CaptureIterableDataset` to record sampler states.
+        dl_kwargs["dataset"] = CaptureIterableDataset(dataset=dl_kwargs["dataset"])
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:306]:
```suggestion
        dl_kwargs["dataset"] = CaptureIterableDataset(dataset=dataset)
```

File: pytorch_lightning/utilities/data.py [hunk 8]
```diff
@@ -305,3 +297,15 @@ def _dataloader_init_kwargs_resolve_sampler(
...
+    elif get_len(dataset) != float("inf"):
+        dl_kwargs["dataset"] = CaptureMapDataset(dataset=dl_kwargs["dataset"])
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:308]:
```suggestion
        dl_kwargs["dataset"] = CaptureMapDataset(dataset=dataset)
```

File: pytorch_lightning/utilities/auto_restart.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:434]: Done !

File: pytorch_lightning/utilities/auto_restart.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:440]: You need the `nonlocal num_batches_fetched` or it would be unbounded.

File: pytorch_lightning/utilities/auto_restart.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:433]: Shortly.

File: pytorch_lightning/utilities/auto_restart.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/utilities/auto_restart.py:442]: We are patching the iterator `_next_data`.

```
    iterator._next_data = _next_data_wrapper(
        iterator._next_data, iterator, dataloader, num_batches_fetched, data_fetcher
    )
```

With the refactor, this won't be necessary anymore as we will be subclassing the iterator directly and overriding _next_data.

File: pytorch_lightning/utilities/data.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/utilities/data.py:306]: Done !

File: pytorch_lightning/utilities/data.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/utilities/data.py:308]: Done !
